From 8a4eb886c659da78bc88432dbce93c6b0cc8ff1b Mon Sep 17 00:00:00 2001
From: Andrii Koval <andrii.koval@lge.com>
Date: Wed, 29 Jul 2015 11:26:34 +0300
Subject: [PATCH] Optimize V8's JSON parser (second part)

:Release Notes:
Implemented second optimization point to improve memory usage and performance
of JSON parser.

:Detailed Notes:
Main idea for optimization. Add Buffer Manager which allocates string buffer
once for one JSON parser instance and reuses it for another strings. It useful
for SlowScanJsonString() which at the beginning allocates buffer for the input
string convertation.
Notes: If requested buffer bigger than existing, buffer will be reallocated.
Manager will allocate buffer only if parser requests it, and will deallocate
it when JSON parser be destroyed.
Finally, this optimization doesn't work without previos optimization
([DRD-6249]).

:Testing Performed:
Locally performed the original Node.js unit test (make test)
On ..... performed test with parsing different JSON files

:QA Notes:

:Issues Addressed:
[DRD-7031] - Stabilizing the "second" optimization point for the JSON parser.

Change-Id: I37eaf92da2e90d4e811a1ba716385da162ca7d98
Reviewed-on: https://gpro.lgsvl.com/115025
Reviewed-by: Andrii Koval <andrii.koval@lge.com>
Tested-by: Andrii Koval <andrii.koval@lge.com>
Reviewed-by: Sergii Nikolaiev <sergii.nikolaiev@lge.com>
Reviewed-by: Tigran Avanesov <tigran.avanesov@lge.com>
---
 deps/v8/src/json-parser.cc | 14 ++++---
 deps/v8/src/json-parser.h  | 95 ++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 103 insertions(+), 6 deletions(-)

diff --git a/deps/v8/src/json-parser.cc b/deps/v8/src/json-parser.cc
index efba265..65410db 100644
--- a/deps/v8/src/json-parser.cc
+++ b/deps/v8/src/json-parser.cc
@@ -142,7 +142,8 @@ JsonParser<seq_one_byte>::JsonParser(Isolate* isolate, Handle<String> source)
       object_constructor_(isolate_->native_context()->object_function(),
                           isolate_),
       position_(-1),
-      properties_(&zone_) {
+      properties_(&zone_),
+      buf_manager_(isolate_) {
   source_ = String::Flatten(source_);
   pretenure_ = (source_length_ >= kPretenureTreshold) ? TENURED : NOT_TENURED;
 
@@ -726,8 +727,7 @@ Handle<String> JsonParser<seq_one_byte>::SlowScanJsonString(
   int string_end_position = FindEndOfStr(position_);
   int itput_str_length = string_end_position - position_ + count;
   int length = Min(max_length, itput_str_length);
-  Handle<StringType> seq_string =
-      NewRawString<StringType>(factory(), length, pretenure_);
+  Handle<StringType> seq_string = buf_manager_.GetBufString<StringType>(length);
   // Copy prefix into seq_str.
   SinkChar* dest = seq_string->GetChars();
   String::WriteToFlat(*prefix, dest, start, end);
@@ -737,7 +737,8 @@ Handle<String> JsonParser<seq_one_byte>::SlowScanJsonString(
     if (c0_ < 0x20) return Handle<String>::null();
     if (count >= length) {
       // We need to create a longer sequential string for the result.
-      return SlowScanJsonString<StringType, SinkChar>(seq_string, 0, count);
+      Handle<String> processed_str = buf_manager_.GetRealStrFromBuf<StringType>(count);
+      return SlowScanJsonString<StringType, SinkChar>(processed_str, 0, count);
     }
     if (c0_ != '\\') {
       // If the sink can contain UC16 characters, or source_ contains only
@@ -794,7 +795,8 @@ Handle<String> JsonParser<seq_one_byte>::SlowScanJsonString(
             // char.
             position_ -= 6;  // Rewind position_ to \ in \uxxxx.
             Advance();
-            return SlowScanJsonString<SeqTwoByteString, uc16>(seq_string, 0,
+            Handle<String> processed_str = buf_manager_.GetRealStrFromBuf<StringType>(count);
+            return SlowScanJsonString<SeqTwoByteString, uc16>(processed_str, 0,
                                                               count);
           }
         }
@@ -810,7 +812,7 @@ Handle<String> JsonParser<seq_one_byte>::SlowScanJsonString(
   AdvanceSkipWhitespace();
 
   // Shrink seq_string length to count and return.
-  return SeqString::Truncate(seq_string, count);
+  return buf_manager_.GetRealStrFromBuf<StringType>(count);
 }
 
 template <bool seq_one_byte>
diff --git a/deps/v8/src/json-parser.h b/deps/v8/src/json-parser.h
index 27d33a3..39f9d92 100644
--- a/deps/v8/src/json-parser.h
+++ b/deps/v8/src/json-parser.h
@@ -10,9 +10,102 @@
 #include "src/objects.h"
 #include "src/zone/zone-containers.h"
 
+#include <memory>
+
 namespace v8 {
 namespace internal {
 
+namespace
+{
+//Buffer manager for SlowScanStr functionality
+class SlowScanBufManager
+{
+public:
+  SlowScanBufManager(Isolate* isolate)
+  : isolate_(isolate)
+  , factory_(isolate_->factory())
+  , pretenure_(NOT_TENURED)
+  {
+  }
+
+  ~SlowScanBufManager() {
+  }
+
+  void SetPretenure(PretenureFlag pretenure) {
+    pretenure_ = pretenure;
+  }
+
+  template <typename StringType>
+  Handle<StringType> GetBufString(int length);
+
+  template <typename StringType>
+  Handle<StringType> GetRealStrFromBuf(int length);
+
+private:
+  Isolate* isolate_;
+  Factory* factory_;
+  PretenureFlag pretenure_;
+
+  Handle<SeqTwoByteString> tb_str_;
+  Handle<SeqOneByteString> ob_str_;
+
+  std::auto_ptr<DeferredHandles> tb_deferred_handles_;
+  std::auto_ptr<DeferredHandles> ob_deferred_handles_;
+
+private:
+  SlowScanBufManager(const SlowScanBufManager&);
+  SlowScanBufManager& operator=(const SlowScanBufManager&);
+};
+
+template <>
+inline Handle<SeqTwoByteString> SlowScanBufManager::GetBufString(int length) {
+  if (tb_str_.is_null() || tb_str_->length() < length) {
+    DeferredHandleScope deferred_hs(isolate_);
+    tb_str_ = factory_->NewRawTwoByteString(length, pretenure_).ToHandleChecked();
+    tb_deferred_handles_.reset(deferred_hs.Detach());
+  }
+
+  return tb_str_;
+}
+
+template <>
+inline Handle<SeqOneByteString> SlowScanBufManager::GetBufString(int length) {
+  if (ob_str_.is_null() || ob_str_->length() < length){
+    DeferredHandleScope deferred_hs(isolate_);
+    ob_str_ = factory_->NewRawOneByteString(length, pretenure_).ToHandleChecked();
+    ob_deferred_handles_.reset(deferred_hs.Detach());
+  }
+
+  return ob_str_;
+}
+
+template <>
+Handle<SeqTwoByteString> SlowScanBufManager::GetRealStrFromBuf(int length)
+{
+  Handle<SeqTwoByteString> seq_str = factory_->NewRawTwoByteString(length, pretenure_).ToHandleChecked();
+
+  uc16* tmp_ptr = tb_str_->GetChars();
+  for (int i = 0; i < length; ++i) {
+    seq_str->SeqTwoByteStringSet(i, tmp_ptr[i]);
+  }
+
+  return seq_str;
+}
+
+template <>
+Handle<SeqOneByteString> SlowScanBufManager::GetRealStrFromBuf(int length)
+{
+  Handle<SeqOneByteString> seq_str = factory_->NewRawOneByteString(length, pretenure_).ToHandleChecked();
+
+  uint8_t* tmp_ptr = ob_str_->GetChars();
+  for (int i = 0; i < length; ++i) {
+    seq_str->SeqOneByteStringSet(i, tmp_ptr[i]);
+  }
+
+  return seq_str;
+}
+}
+
 enum ParseElementResult { kElementFound, kElementNotFound, kNullHandle };
 
 class JsonParseInternalizer BASE_EMBEDDED {
@@ -187,6 +280,8 @@ class JsonParser BASE_EMBEDDED {
 
   // Property handles are stored here inside ParseJsonObject.
   ZoneVector<Handle<Object>> properties_;
+
+  SlowScanBufManager buf_manager_;
 };
 
 }  // namespace internal
